### PR TITLE
Form

### PR DESCRIPTION
**Pull request by tsawan**
Added form plugin and docs. added a simple reactive form in integration app.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -80,6 +85,7 @@
     }
   },
   "dependencies": {
+    "@angular/forms": "^5.2.9",
```
Comment (amcdnl) [inline package.json]: We would want this to be a peer dependency.

File: src/plugins/form/directive.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    private _formGroupDirective: FormGroupDirective,
+    private _cd: ChangeDetectorRef
+  ) {}
+
+  ngOnInit() {
+    this._store
+      .select(state => getValue(state, `${this.path}.model`))
+      .pipe(takeUntil(this._destroy$))
+      .subscribe(model => {
+        //        if (!this._updating) {
```
Comment (amcdnl) [inline src/plugins/form/directive.ts]: This updating flag is here to prevent circular loops. Why did you remove it?

File: src/plugins/form/directive.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            this._cd.markForCheck();
+          }
+        }
+      });
+
+    this._formGroupDirective.valueChanges
+      .pipe(debounceTime(this.debounce), takeUntil(this._destroy$))
+      .subscribe(value => {
+        //        this._updating = true;
+        this._store.dispatch(
```
Comment (amcdnl) [inline src/plugins/form/directive.ts]: Lets combine these 3 actions into 1.

File: src/plugins/form/form.plugin.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+@Injectable()
+export class NgxsFormPlugin implements NgxsPlugin {
+  constructor() {}
+
+  handle(state, event, next) {
+    const type = getTypeFromInstance(event);
+
+    let nextState = state;
+
+    if (type === "UpdateFormValue" || type === "UpdateForm") {
```
Comment (amcdnl) [inline src/plugins/form/form.plugin.ts]: Use `'` vs `"`

File: src/plugins/form/form.plugin.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      nextState = setValue(nextState, `${event.payload.path}.model`, {
+        ...event.payload.value
+      });
+    }
+
+    if (type === "UpdateFormStatus" || type === "UpdateForm") {
+      nextState = setValue(
+        nextState,
+        `${event.payload.path}.status`,
```
Comment (amcdnl) [inline src/plugins/form/form.plugin.ts]: I'm not sure how this is working.

File: src/plugins/form/form.plugin.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    if (type === "UpdateFormErrors" || type === "UpdateForm") {
+      nextState = setValue(nextState, `${event.payload.path}.errors`, {
+        ...event.payload.errors
+      });
+    }
+
+    if (type === "UpdateFormDirty" || type === "UpdateForm") {
+      nextState = setValue(
+        nextState,
+        `${event.payload.path}.dirty`,
```
Comment (amcdnl) [inline src/plugins/form/form.plugin.ts]: Not sure how this is working.

File: src/plugins/form/directive.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+import { debounceTime } from "rxjs/operators/debounceTime";
+import {
+  UpdateFormStatus,
+  UpdateFormValue,
+  UpdateFormDirty,
+  UpdateFormErrors,
+  UpdateForm
+} from "./actions";
+
+const getValue = (obj, prop) =>
```
Comment (amcdnl) [inline src/plugins/form/directive.ts]: This is available in `./internals` file.

File: src/index.ts [hunk 8]
```diff
@@ -1,13 +1,24 @@
-export { NgxsModule } from './module';
-export { Action } from './action';
-export { Store } from './store';
-export { State } from './state';
-export { Select } from './select';
-export { Actions } from './actions-stream';
-export { ofAction } from './of-action';
-export { NgxsPlugin, NgxsPluginFn, StateContext } from './symbols';
-export { Selector } from './selector';
+export { NgxsModule } from "./module";
```
Comment (amcdnl) [inline src/index.ts]: Should be single quotes.

**Comment by amcdnl** [comment]
@deebloo - With the introduction of this and its new peer depedencies I really feel llke its time to breakout the plugins.

**Comment by yarrgh** [comment]
Each plugin should have it's own npm package to keep unwanted peer dependencies out of apps that don't use them. This plugin could be something like `@ngxs/form-plugin`

File: src/plugins/form/directive.ts (hunk 2 above)
Comment (tsawan) [reply src/plugins/form/directive.ts]: This was causing an issue when I update a form's state in an action and it was not reflecting back in the form. After I commented this code, the issue was resolved and I don't see any circular loops.

File: src/index.ts (hunk 8 above)
Comment (tsawan) [reply src/index.ts]: Done.

File: src/plugins/form/directive.ts (hunk 7 above)
Comment (tsawan) [reply src/plugins/form/directive.ts]: Done.

File: src/plugins/form/form.plugin.ts (hunk 4 above)
Comment (tsawan) [reply src/plugins/form/form.plugin.ts]: Done.

File: src/plugins/form/directive.ts (hunk 3 above)
Comment (tsawan) [reply src/plugins/form/directive.ts]: Done.

File: package.json (hunk 1 above)
Comment (tsawan) [reply package.json]: Done.

File: src/plugins/form/form.plugin.ts (hunk 6 above)
Comment (tsawan) [reply src/plugins/form/form.plugin.ts]: this was the reducer code in ngrx-form. It looks fine to me but let me know if it can be improved?

**Comment by deebloo** [comment]
I agree

**Comment by amcdnl** [comment]
Ya, plan is to break these out into their own packages. @leon and I were talking like this:

```
/modules
   /store
      /src
      /tests
      package.json
  /storage-plugin
    src/
    tests/
    package.json
```

and so on.

**Comment by deebloo** [comment]
@amcdnl @leon I agree. we have a similar setup we use at work. https://github.com/deebloo/angular-multi-lib-starter

**Comment by deebloo** [comment]
@tsawan we just migrated the repo pretty significantly. can you take a look at what is currently on master and submit your PR there? thanks!

**Comment by tsawan** [comment]
@deebloo just to be sure, you want me to fork the current master, apply my changes and submit a fresh PR?

**Comment by deebloo** [comment]
If that is what is easiest yes
